#This script can run on RStudio
#setwd to where the script is located

#Library
library(dplyr)
library(data.table)

#Data path for proj dir
data.path <- "../../../data/PhenoPRS"

#Also load in path to scratch, proj_dir and fgfp data
source("../../parameters/scratch_path.txt")
source("../../parameters/base_dir.R")
source(file.path(pdir, "scripts/parameters/parafile.txt"))

#To merge these we also need the linker file that dave has made
### Load in linker File
#Note: clincal_data_available == 1 does not reduce the n down any further
linker_data <- read.table(file.path(bdir, linker), 
                          header = TRUE, as.is = TRUE, sep = "\t")
complete_data <- dplyr::filter(linker_data, micobiome_data_available == 1 & genotypedata_good_for_gwas == 1
                               & clinical_data_available == 1)

#Read in the phenotype data from the previous script
phenos4prs <- read.csv(file = file.path(data.path, "data_out/fgfp_phenotypes4prs.csv"), header = T)

#Read in covariates for regression
PCs <- read.table(file.path(bdir, genotypePCs), head=T, sep = "\t")

#Edit the name slightly - seemed to be missing a 0 in its linker name so couldnt merge
#sample_name_edit2 is presumed correct as no missingness when merge with this name
batch_data <- fread(file.path(bdir, microbiome_batch_data), header = TRUE, sep = "\t")
batch_data$sample_name_edit <- gsub( "\\.", "_", batch_data$SAMPLE_NAME)
batch_data$sample_name_edit2 <- gsub( "\\.", ".0", batch_data$SAMPLE_NAME)




#Make covariate matrix with the outcomes we are interested in
cov_matrix <- merge(complete_data[,1:4], phenos4prs, by="fgfp_id", all = T) %>%
  merge(PCs[,c(2,4:14)], by.x="fgfp_id", by.y = "ID_2", all = T) %>%
  merge(batch_data[,c(2:7, 11)], by.x="fgfp_id", by.y = "sample_name_edit2", all=T)

#Remove duplicated data
cov_matrix <- cov_matrix[!duplicated(cov_matrix),]

#Change date variables into years
cov_matrix$Aliquote.date_year <- strsplit(cov_matrix$Aliquote.date, "/") %>% sapply(FUN = function(x) x[[3]]) %>% as.numeric()
cov_matrix$Date.of.extraction_year <- strsplit(cov_matrix$Date.of.extraction, "/") %>% sapply(FUN = function(x) x[[3]]) %>% as.numeric()


#85% were cut and so fill in missing with cut
cov_matrix <- cov_matrix %>%
  mutate(drilled.cut = ifelse(drilled.cut == "", "cut", drilled.cut))


#Load in the prs traits that we have made
load(file.path(data.path, "phenosumstats/filtered_sumstats_list.RData"))

#And specify the function for generating the file names
#Specify a function for cleaning directory names
clean_directory_name <- function(x) {
  x %>%
    stringr::str_replace_all("[^[:alnum:]]+", "_") %>% 
    stringr::str_replace_all("_+", "_")   
}

prs_names <- clean_directory_name(names(snps_df))

#Read in our prs scores for each trait
prs <- list()
for(trait in prs_names){
  PRSice_file <- file.path(data.path, "phenosumstats", trait, "prsice_out", paste0(trait, ".all_score"))
  if(file.exists(PRSice_file)){
    prs[[trait]] <- read.table(file = PRSice_file, header = T)
  }else{prs[[trait]] <- NA}
}

#Two traits removed as no snps to generate the PRS!

#Also add BMI as was generated by OpenGWAS
prs[["BMI"]] <- read.table(file = file.path(data.path, "data_out/old/prs_out/BMI/prsice_out/BMI.all_score"), header = T)
  
#Also add triglyceride from old as one generate from GWASCatalog seems to not be working properly
prs[["triglycerides"]] <- read.table(file = file.path(data.path, "data_out/old/prs_out/triglycerides/prsice_out/triglycerides.all_score"), header = T)

#Now generate a table from this list
prs_df <- do.call(cbind, prs) %>% dplyr::select(-ends_with("FID"), -ends_with("IID"), Alanine_aminotransferase_levels.IID) %>%
  rename(IID = Alanine_aminotransferase_levels.IID)

pheno_covariate_prs <- merge(prs_df, linker_data, by.x = "IID", by.y = "genetic_id", all = T) %>%
  merge(cov_matrix, by = "fgfp_id", all = T) 


### Lets also load in the fgfp gwased traits
#Load in microbial data
gwased_mt_data <- fread(file.path(bdir, microbiome_GWAS_source), 
                        header = F, sep = "\t")
header <- c("linker", fread(cmd = paste("head -n 1", file.path(bdir, microbiome_GWAS_source))) %>% colnames())
colnames(gwased_mt_data) <- header



#Now lets save the clean dataframe / list into the data directory
#Some repeats of data so have used distinct() command to make sure these are removed
data4prs <- list(pheno_covariate_prs = distinct(pheno_covariate_prs), 
                 gwasedmts = gwased_mt_data)

prs_names_rm <- names(prs)[!names(prs) %in% c("Inflammatory_bowel_disease_SPA_correction_", "Type_milk_consumed_soya_with_calcium")]

data4prs[["matchedvars"]] <- data.frame(prs = prs_names_rm,
                                  pheno = c("GPT_UL",
                                            "beer_comsuption",
                                            "coffee_last_consumption_before_sampling",
                                            "CK_UL",
                                            "J01CA04",
                                            "e.GFR",
                                            "G03CA04_CC06",
                                            "fruits_last_consumption_before_sampling",
                                            "Gamma.GT_UL",
                                            "HDL.chol_mgdL",
                                            "Hemoglobine_gdL",
                                            "hip_circumference",
                                            "ibs",
                                            "MCHC_g_dL",
                                            "R06AX28",
                                            "L04AX01",
                                            "average_cigarettes_consumption_per_day_last3months",
                                            "meat_average_consumption_last_week",
                                            "RBC_milj_mm3",
                                            "Creatinine_mgdL",
                                            "Ureum_mgdL",
                                            "Urinezuur_mgdL",
                                            "sleeping_hours_per_day",
                                            "have_smoked_last3month",
                                            "Height",
                                            "stoole_score",
                                            "Triglyceriden_mgdL",
                                            "had_ulcerative_colitis",
                                            "BMI"))

save(data4prs, file = file.path(data.path, "data_out/fgfpdata4prs.RData"))
