---
title: "h2_mts Results"
author: "Alec McKinlay"
date: "2024-07-02"
output:
  html_document:
---

```{r, echo = FALSE, message=FALSE, warning=FALSE}

#Set options for the fig sides
knitr::opts_chunk$set(fig.width=14, fig.height=8, echo = F, message=F, warning=F)

#Set up and read in files
#setwd to 05_Markdowns folder

#Library 
library(tidyverse)
library(data.table)
library(readxl)
library(kableExtra)
library(ggplot2)
library(ggforestplot)
library(DataExplorer)
library(patchwork)
library(forcats)
library(makepipe)

#Data path for proj dir
data.path <- "../../../data/PhenoPRS"

#Also load in path to scratch, proj_dir and fgfp data
source("../../parameters/base_dir.R")

#Load GWAS table in
gwas_table <- read.csv(file = file.path(data.path, "ProcessedGWASTable_OG_GC.csv"), header = T)

#Load the data in, incase we use this
load(file.path(data.path, "data_out/fgfpdata4prs.RData"))

#Load in the data from RNT regression script (03.1), first we can check if needs to be updated before loading
#make_with_source(
#  source = "../03_Regressions/03.1_RegRNTs.R", 
#  targets = file.path(data.path, "data_out/regout_covarprs.RData"), 
#  dependencies = c(file.path(data.path, "ProcessedGWASTable_OG_GC.csv"),
#                   file.path(data.path, "data_out/fgfpdata4prs.RData"))
#)

#Update so that they load as one object rather than individual ones

#Now load the output of reg1
load(file.path(data.path, "data_out/regout_covarprs.RData"))

#Out output of reg2
load(file.path(data.path, "data_out/regout_prsRNTs.RData"))

#And the observational estimates
load(file.path(data.path, "data_out/regout_observational.RData"))

#State filepath to our figures in the markdown
fallonyjpeg <- file.path(data.path, "markdowns/figs/fallonyfig.jpeg")
methodsjpeg <- file.path(data.path, "markdowns/figs/method.jpg")


```

## Are microbial covariates driving associations between the host genome and Gut bacterial traits?

* Recent literature aimed at quantifying covariates of the microbiome i.e. what phenotypes drive variation?
* Could the hertiability observed for MTs be driven by these covairates of the microbiome?

### Methods

1.  Select covariates of the microbiome
2.  Generate PRSs for these covariates
3.  Regress PRSs onto those heritable MTs from Dave's paper, do we see signal?
4.  For those with signal, we can then run further analyses to unpick these associations:
    (a) Recalculation of h2 of heritable MTs before and after PRS adjustement
    (b) Observational estimate between the MT and covariate
    (c) Genetic correlation between the two traits
5.  For those with consistent signal for our methodologies, we can generate a PRS of the microbial trait and run a parallel 2SMR analysis with the covariable and microbial trait

##### Methodological overview

![Methodological overview Figure](`r methodsjpeg`)

### Decide on covariates

Fallony et al studies which covariates in FGFP explained the most variation in beta diversity. We have taken the top covariates forward to generate PRSs for

##### Fallony covariates

![Fallony Figure](`r fallonyjpeg`)

### Generate PRS for covariates

The following traits have been chosen to generate PRS for in FGFP. These were selected through GWAS Catalog / OpenGWAS, filtering to EUR only with largest sample sizes and number of SNPs surpassing genome-wide sig threhsold.

```{r}
rm_traits <- c("Bristol Stool Score", "creatinine3")
gwas_table_filt <- gwas_table[,c(1:2,5,7, 8)] %>% dplyr::filter(!Falony.trait %in% rm_traits)
gwas_table_filt %>%
  kbl() %>%
  kable_styling()
```

* A total of `r nrow(gwas_table_filt)` traits were chosen where PRSs were generated
* Now we can generate a forest plot to see if these PRSs associate to the trait, a sanity check. Many of these variables are not continuous, we have started with just those that are continuous for ease.

#### Covariate Phenotype distribution and missingness in FGFP

Lets look at how our phenotypes are distributed:

```{r}

#We are only interested in individuals from FGFP with pheno / geno / micro
fgfp_data <- filter(data4prs$pheno_covariate_prs, !is.na(hemoglobin.Pt_5e.08) |
                      micobiome_data_available == 1 | clinical_data_available == 1)

fgfp_data_cc <- filter(fgfp_data, !is.na(hemoglobin.Pt_5e.08) &
                      micobiome_data_available == 1 & clinical_data_available == 1)
```

* In our FGFP dataframe we data for `r nrow(fgfp_data)` participants with genotype OR clinic OR microbiome data.
* There are `r nrow(fgfp_data_cc)` participants with genotype AND clinic AND microbiome data

Lets look at the distribution of our phenotypes and the missingness structure so we can decide how best to model them for subsequent analyses:

```{r}
#Lets look at the distribution of our variables
plot_histogram(fgfp_data[,regout_covarprs$data4reg$pheno])
```

#### PRS -\> Covar (sanity check)

Now we can see if our PRSs associate with the covariate they are trained to capture:

```{r}
#Format p values so can be plotted
regout_covarprs$regout_covarprs_cont_scale$pvalue_label = formatC(regout_covarprs$regout_covarprs_cont_scale$p.value, format = "e", digits = 2)

regout_covarprs$regout_covarprs_cont_scale %>% ggplot(aes(x = fct_rev(term), y = std_estimate, ymin = conf.low, ymax = conf.high)) +
  geom_pointrange(color = "black", size = 0.5) +
  geom_hline(yintercept = 0, color = "steelblue") +
  geom_text(aes(label = pvalue_label), color = "red", hjust = -0.2, size = 3) +
  coord_flip() +
  xlab("Phenotype") +
  ylab("Beta Coefficient (95% Confidence Interval)") +
  labs(title = "Linear Regression Models Estimating the PRS of the covariate on the covariate") 
```

The plot shows beta coefficents of PRS of covariate regressed onto the phenotype value of covariate in FGFP. Red numbers indicate pvalues. CIs are extremely narrow hence looks like they have been omitted from the plot.

We can see for creatinine we have a negative association. This is because our PRS is proxying kidney function e.g. estimated GFR. We also have this variable measured in FGFP however it has a high degree if missingness, hence we have used creatinine in its place

This missingness of the two variables can be observed below. The table shows the % missingness of the variables.

```{r}
miss <- cbind(fgfp_data[,c("e.GFR", "Creatinine_mgdL")] %>%
  summarise_all(~ mean(is.na(.)) * 100) %>%
  gather(variable, missing_percent), colSums(!is.na(fgfp_data[,c("e.GFR", "Creatinine_mgdL")])))

colnames(miss)[3] <- "n_obs"

miss[,-1] %>%
  kbl() %>%
  kable_styling()
```

Due to the missingness structure, we will take forward the creatinine variable to proxy kindey function instead.

### PRS -> MTs
* First we can focus our analyses on Dave's RNT Microbial traits which were generated for the FGFP GWAS

Below highlight the MT traits which showed evidence of h2 as estimated by GREML:

```{r}
#Read in hughes supp file with h2 results
hughes_h2 <- read_xlsx(path = file.path(data.path, "Hughes_mtGWAS.xlsx"),
                       sheet = "Table S3",
                       skip = 4)

#Change pval cols to numeric so we can filter on these values
pvalcols <- colnames(hughes_h2)[colnames(hughes_h2) %>% endsWith("_pval")]
hughes_h2[,pvalcols] <- sapply(hughes_h2[,pvalcols], as.numeric)

#Now find out which RNT traits showed evidence for heritability
RNT_h2_traits <- filter(hughes_h2, RNT_pval < 0.05) %>% dplyr::select(TaxaName, RNT_h2, RNT_se, RNT_pval, RNT_n)
RNT_h2_traits %>%
  kbl(digits = 2) %>%
  kable_styling()

```

* Now we regress our PRSs onto these heritable microbial traits to see where there is evidence of signal

```{r}
#Lets filter our scaled lm output into only those heritable bugs
filterandnamechange <- function(x){
df <- filter(x, mt %in% paste0(RNT_h2_traits$TaxaName, "_RNTRes"))
df$PRS <- gsub(".Pt_5e.08", "_PRS", df$term) 
return(df)
}

h2mtsres <- lapply(regout_prsRNTs, 
                 FUN = filterandnamechange)

signal <- lapply(h2mtsres, 
                 function(x) filter(x, p.value < 0.05))

#What prs' are we looking at?
prs2use <- regout_covarprs$regout_covarprs_cont_scale$`data4reg[data4reg$pheno_vartype == "cont", "prs"]` %>% unique()
```

* `r nrow(RNT_h2_traits)` Microbial traits and `r length(prs2use)` covariate PRSs have been used for the regressions
* All of the regressions run can be visualised below:

```{r}
ggforestplot::forestplot(
  df = h2mtsres$regout_RNTs_univar,
  name = mt,
  estimate = estimate,
  se = std.error,
  pvalue = p.value,
  psignif = 0.05,
  colour = PRS,
  xlab = "SD increase in Rank normal transformed Microbial trait per 1-SD increase in polygenic loading to the covariates of interest",
  title = "Regression coefficents from PRS covariate on Microbial trait")
```

* Lets filter to those beta coefficents with evidence of signal
* And also present those with bonferonni corrected signal, those points which are filled in show evidence of bonferroni correct association

```{r}
ggforestplot::forestplot(
  df = signal$regout_RNTs_univar,
  name = mt,
  estimate = estimate,
  se = std.error,
  pvalue = p.value,
  psignif = 0.05/length(prs2use),
  colour = PRS,
  xlab = "SD increase in Rank normal transformed Microbial trait per 1-SD increase in polygenic loading to the covariates of interest",
  title = "Regression coefficents from PRS covariate on Microbial trait")
```

### Observational Esimate: Covariate -> RNT MT

* Do our observational estimates agree with the traits showing evidence for an association with the PRS?

```{r}
obs_est$PRS <- rep("Observational", nrow(obs_est))
colnames(obs_est)[6] <- "nobs(lm2)"
obs_sigprs <- rbind(signal$regout_RNTs_univar, obs_est)

ggforestplot::forestplot(
  df = obs_sigprs,
  name = mt,
  estimate = estimate,
  se = std.error,
  pvalue = p.value,
  psignif = 0.05/length(prs2use),
  colour = PRS,
  xlab = "SD increase in Rank normal transformed Microbial trait per 1-SD increase in polygenic loading of the covariate / the measure covariate",
  title = "Regression coefficents from PRS covariate on Microbial trait")
```

* We see consistency in for the latter trait
* Do we also see this consistency in Genetic correlation between the two observational triats?

## Genetic Correlation