---
title: "h2_mts Results"
author: "Alec McKinlay"
date: "2024-07-02"
output:
  html_document:
---

```{r, echo = FALSE, message=FALSE, warning=FALSE}

#Set options for the fig sides
knitr::opts_chunk$set(fig.width=14, fig.height=8, echo = F, message=F, warning=F)

#Set up and read in files
#setwd to 05_Markdowns folder

#Library 
library(tidyverse)
library(data.table)
library(readxl)
library(kableExtra)
library(ggplot2)
library(ggforestplot)
library(DataExplorer)
library(patchwork)
library(forcats)
library(makepipe)
library(tidyr)
library(ggh4x)

#Data path for proj dir
data.path <- "../../../data/PhenoPRS"

#Also load in path to scratch, proj_dir and fgfp data
source("../../parameters/base_dir.R")

#Load GWAS table in
gwas_table <- read.csv(file = file.path(data.path, "ProcessedGWASTable_OG_GC.csv"), header = T)

#Load the data in, incase we use this
load(file.path(data.path, "data_out/fgfpdata4prs.RData"))

#Load in the data from RNT regression script (03.1), first we can check if needs to be updated before loading
make_with_source(
  source = "../03_Regressions/03.1_RegRNTs.R", 
  targets = file.path(data.path, "data_out/data_Reg03.1.RData"), 
  dependencies = c(file.path(data.path, "ProcessedGWASTable_OG_GC.csv"),
                   file.path(data.path, "data_out/fgfpdata4prs.RData"))
)

#Now load the data
load(file.path(data.path, "data_out/data_Reg03.1.RData"))

#Load in the consistent associations from script 03.2
make_with_source(
  source = "../03_Regressions/03.2_FindConsistentAssociations.R", 
  targets = file.path(data.path, "data_out/data_ConsistentAssoc03.2.RData"), 
  dependencies = file.path(data.path, "data_out/data_Reg03.1.RData")
)

#Now load the data
load(file.path(data.path, "data_out/data_ConsistentAssoc03.2.RData"))

#Now load in the GREML genetic correlation results
#Can't get this make_with_source to run - go back and fix this
#make_with_source(
#  source = "../04_HeritabilityCalcs/04.3_GeneticCor/04.3.3_CompileCorEst.R", 
#  targets = file.path(data.path, "greml/greml_out/h2out_cor_dircons/clean_out.txt"), 
#  dependencies = file.path(data.path, "greml/greml_out/h2out_cor_dircons")
#)

genetic_cor <- read.table(file.path(data.path, "greml/greml_out/h2out_cor_dircons/clean_out.txt"), header = T,
                          sep = "\t")

obs_prs_cor_filt <- read.table(file = file.path(data.path, "greml/greml_out/h2out_cor_dircons/clean_out_withobsprs.txt"), header = T, sep = "\t")



#Now load in the heritability adjustment results
h2out <- list()
for(i in c("dircons", "h2bugs")){
  h2out[[i]] <- read.table(file.path(data.path, paste0("greml/greml_out/h2_out/", i, "/cleanh2est.txt")), header = T, sep = "\t")
}

#State filepath to our figures in the markdown
fallonyjpeg <- file.path(data.path, "markdowns/figs/fallonyfig.jpeg")
methodsjpeg <- file.path(data.path, "markdowns/figs/method.jpg")


```

## Are microbial covariates driving associations between the host genome and Gut bacterial traits?

* Recent literature aimed at quantifying covariates of the microbiome i.e. what phenotypes drive variation?
* Could the hertiability observed for MTs be driven by these covairates of the microbiome?

### Methods

1.  Select covariates of the microbiome
2.  Generate observational estimate for convariate on microbial trait
3.  Generate PRSs for these covariates  
4.  For those with observational signal, we can then run further analyses to unpick these associations:
    (a) Regress PRSs onto Microbial trait
    (b) Genetic correlation between the two traits
    (c) Recalculation of h2 of heritable MTs before and after PRS adjustement - is the covar driving mt h2?
5.  For those with directionally consistent signal for all steps, we can then generate a PRS of the microbial trait and the covariate to run a parallel 2SMR analysis with an outcome of interest. Preferably an outcome where we know to be a strong causal association between the covariate and the outcome.

##### Methodological overview

![Methodological overview Figure](`r methodsjpeg`)

### Decide on covariates

Fallony et al studies which covariates in FGFP explained the most variation in beta diversity. We have taken the top covariates forward to generate PRSs for

##### Fallony covariates

![Fallony Figure](`r fallonyjpeg`)

### Generate PRS for covariates

The following traits have been chosen to generate PRS for in FGFP. These were selected through GWAS Catalog / OpenGWAS, filtering to EUR only with largest sample sizes and number of SNPs surpassing genome-wide sig threhsold.

```{r}
rm_traits <- c("Bristol Stool Score", "creatinine3")
gwas_table_filt <- gwas_table[,c(1:2,5,7, 8)] %>% dplyr::filter(!Falony.trait %in% rm_traits)
gwas_table_filt %>%
  kbl() %>%
  kable_styling()
```

* A total of `r nrow(gwas_table_filt)` traits were chosen where PRSs were generated
* Now we can generate a forest plot to see if these PRSs associate to the trait, a sanity check. Many of these variables are not continuous, we have started with just those that are continuous for ease.

#### Covariate Phenotype distribution and missingness in FGFP

Lets look at how our phenotypes are distributed:

```{r}

#We are only interested in individuals from FGFP with pheno / geno / micro
fgfp_data <- filter(data4prs$pheno_covariate_prs, !is.na(hemoglobin.Pt_5e.08) |
                      micobiome_data_available == 1 | clinical_data_available == 1)

fgfp_data_cc <- filter(fgfp_data, !is.na(hemoglobin.Pt_5e.08) &
                      micobiome_data_available == 1 & clinical_data_available == 1)
```

* In our FGFP dataframe we data for `r nrow(fgfp_data)` participants with genotype OR clinic OR microbiome data.
* There are `r nrow(fgfp_data_cc)` participants with genotype AND clinic AND microbiome data

Lets look at the distribution of our phenotypes and the missingness structure so we can decide how best to model them for subsequent analyses:

```{r}
#Lets look at the distribution of our variables
plot_histogram(fgfp_data[,regout_covarprs$data4reg$pheno])
```

#### PRS -\> Covar (sanity check)

Now we can see if our PRSs associate with the covariate they are trained to capture:

```{r}
#Format p values so can be plotted left this in if wanted to add back into the plot
#regout_covarprs$regout_covarprs_cont_scale$pvalue_label = formatC(regout_covarprs$regout_covarprs_cont_scale$p.value, format = "e", digits = 2)

regout_covarprs$regout_covarprs_cont_scale %>% ggplot(aes(x = fct_rev(term), y = estimate, ymin = conf.low, ymax = conf.high)) +
  geom_pointrange(color = "black", size = 0.5) +
  geom_hline(yintercept = 0, color = "steelblue") +
  #geom_text(aes(label = pvalue_label), color = "red", hjust = -0.2, size = 3) +
  coord_flip() +
  xlab("Phenotype") +
  ylab("Standardised Beta Coefficient and 95% Confidence Interval showing SD increase in phenotype per 1-SD increase in polygenic loading to the phenotype") +
  labs(title = "Linear Regression Models Estimating the PRS of the covariate on the covariate") 
```

The plot shows beta coefficents of PRS of covariate regressed onto the phenotype value of covariate in FGFP. Red numbers indicate pvalues. CIs are extremely narrow hence looks like they have been omitted from the plot.

We can see for creatinine we have a negative association. This is because our PRS is proxying kidney function e.g. estimated GFR. We also have this variable measured in FGFP however it has a high degree if missingness, hence we have used creatinine in its place

This missingness of the two variables can be observed below. The table shows the % missingness of the variables.

```{r}
miss <- cbind(fgfp_data[,c("e.GFR", "Creatinine_mgdL")] %>%
  summarise_all(~ mean(is.na(.)) * 100) %>%
  gather(variable, missing_percent), colSums(!is.na(fgfp_data[,c("e.GFR", "Creatinine_mgdL")])))

colnames(miss)[3] <- "n_obs"

miss[,-1] %>%
  kbl() %>%
  kable_styling()
```

* Due to the missingness structure, we will take forward the creatinine variable to proxy kidney function instead.
* At the moment we have just removed these two variables

### Observational Analysis. Covariate -> RNT MT

* First we want to run our observational analyses to see where we have evidence of signal

```{r, fig.width=18, fig.height=15}
#Remove unwanted obs estimates as not looking at these for now
obs_est_rm <- filter(obs_est, !term %in% c("e.GFR", "Creatinine_mgdL"))

ggforestplot::forestplot(
  df = obs_est_rm,
  name = mt,
  estimate = estimate,
  se = std.error,
  pvalue = p.value,
  colour = covar_name,
  xlab = "SD increase in Rank normal transformed Microbial trait per 1-SD increase of the covariate in FGFP",
  title = "Regression coefficents from Observational analyses of Covariate regressed on RNT Microbial Traits in FGFP") +
  facet_wrap(
    facets = ~group,
    ncol = 3,  # Number of columns
    nrow = 2,  # Number of rows
    scales = "free_y",
  )

#We can also look at those with sig observational results
ggforestplot::forestplot(
  df = filter(obs_est_rm, p.value < 0.05),
  name = mt,
  estimate = estimate,
  se = std.error,
  pvalue = p.value,
  colour = covar_name,
  xlab = "SD increase in Rank normal transformed Microbial trait per 1-SD increase in polygenic loading of the covariate / the measure covariate",
  title = "Regression coefficents from Observational analyses of Covariate regressed on RNT Microbial Traits in FGFP") +
  ggforce::facet_wrap_paginate(
    facets = ~group,
    ncol = 3,  # Number of columns
    nrow = 2,  # Number of rows
    scales = "free_y"
  )

```

* `r obs_est_rm$mt %>% unique %>% length` Microbial traits and `r obs_est_rm$covar_name %>% unique %>% length` covariate PRSs have been used for the regressions. 
* A total of `r nrow(obs_est_rm)` observational regressions
* `r filter(obs_est_rm, p.value < 0.05) %>% nrow` regressions showed a suggestive evidence of an associaiton P < 0.05.
* `r filter(obs_est_rm, p.value < 0.05/nrow(obs_est_rm)) %>% nrow` regressions showed strong evidence of an association (bonferonni corrected: 0.05 / `r nrow(obs_est_rm)`)


### PRS Analysis: Covariate PRS -> RNT MT
* Now we can Generate PRSs for our covariates (as shown above) and regress these PRSs onto the RNT MTs.
* We can further subset our data to those results with suggestive observational results and beta directional agreement in the PRS analysis


```{r, fig.width=18, fig.height=15}
#We can also look at those with sig observational results

#First reshape the data
obsest_prsest_long <- obsest_prsest %>%
  pivot_longer(cols = c(estimate_obs, estimate_prs), 
               names_to = "type", 
               values_to = "estimate") %>%
  pivot_longer(cols = c(std.error_obs, std.error_prs), 
               names_to = "error_type", 
               values_to = "std.error") %>%
  pivot_longer(cols = c(p.value_obs, p.value_prs), 
               names_to = "p_value_type", 
               values_to = "p.value") %>%
  filter((type == "estimate_obs" & error_type == "std.error_obs" & p_value_type == "p.value_obs") |
         (type == "estimate_prs" & error_type == "std.error_prs" & p_value_type == "p.value_prs")) %>%
  mutate(type = recode(type, estimate_obs = "Observational", estimate_prs = "PRS"))



#Forest plot for the observational and PRS estimate
ggforestplot::forestplot(
  df = obsest_prsest_long,
  name = covar_name,
  estimate = estimate,
  se = std.error,
  pvalue = p.value,
  colour = type,
  xlab = "SD increase in Rank normal transformed Microbial trait per 1-SD increase in polygenic loading of the covariate / the measure covariate",
  title = "Regression coefficents from Observational analyses of Covariate regressed on RNT Microbial Traits in FGFP") +
  ggforce::facet_wrap_paginate(
    facets = ~mt,
    ncol = 4,  # Number of columns
    nrow = 15,  # Number of rows
    scales = "free_y"
  )
```


# Genetic Correlation

* For those traits with consistent estimates in PRS and observational analysis, does Genetic Cor estimated through GREML also agree?

```{r}
# Combine the observational and PRS estimates for plotting
plot_data <- obs_prs_cor_filt %>%
  select(term, mt, estimate_obs, std.error_obs, estimate_prs, std.error_prs) %>%
  pivot_longer(cols = c(estimate_obs, estimate_prs),
               names_to = "type",
               values_to = "estimate") %>%
  mutate(std.error = ifelse(type == "estimate_obs", std.error_obs, std.error_prs),
         type = ifelse(type == "estimate_obs", "Observational", "PRS"))

# Create a separate data frame for the genetic correlation estimates
rg_data <- obs_prs_cor_filt %>%
  select(term, mt, rg, rg_se) %>%
  rename(estimate = rg, std.error = rg_se) %>%
  mutate(type = "Genetic Correlation")

# Combine the data frames for plotting
combined_data <- bind_rows(plot_data, rg_data)

# Filter out mts that are not relevant for specific terms
filtered_data <- combined_data %>%
  group_by(term, type) %>%
  filter(!is.na(estimate)) %>%
  ungroup()

# Reorder the type factor to ensure the correct order in the plot
filtered_data$type <- factor(filtered_data$type, levels = c("Observational", "PRS", "Genetic Correlation"))

# Create the forest plot with facets
ggplot(filtered_data, aes(x = estimate, y = mt, color = type)) +
  geom_point(position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = estimate - 1.96 * std.error,
                    xmax = estimate + 1.96 * std.error),
                width = 0.2, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "red") +  # Add dashed red line at x = 0
  facet_grid(term ~ type, scales = "free", switch = "y") +
  theme_minimal() +
  labs(title = "Forest Plot of Observational, PRS Betas, and Genetic Correlation",
       x = "Estimate",
       y = "Microbiome Trait",
       color = "Estimate Type") +
  theme(legend.position = "bottom",
        strip.text.y = element_text(angle = 0),  # Adjust angle for term labels
        panel.grid.minor = element_blank())  # Remove minor grid lines
```


# Adjusted h2

* For those traits with consistent estimates in PRS, observational analysis and genetic correlation, does the h2 of the MT change after adjusting for the covariate PRS?

```{r}
#First extract the bug names by removing any part of the string after "_res_" in the trait column 
h2out[["dircons"]] <- h2out[["dircons"]] %>%
  mutate(bug = gsub("_res_.*", "", trait))

#Plot estimates from h2out[["dircons"]] so we can see the adjusted and unadjusted h2 estimates, if ends with 5e.08, these are the adjusted phenotypes
h2out[["dircons"]] %>%
  ggplot(aes(x = h2, y = bug, color = ifelse(grepl("5e.08", trait), "Adjusted", "Unadjusted"))) +
  geom_point(position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = h2 - 1.96 * se, xmax = h2 + 1.96 * se), position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "red") +
  theme_minimal() +
  labs(title = "Heritability Estimates of Microbiome Traits Before and After Adjustment for Covariate PRS",
       x = "Heritability Estimate",
       y = "Microbiome Trait",
       color = "Adjustment Type") +
  theme(legend.position = "bottom")
```

* It appears no difference in the hertiable estimates. What if instead we took the most heritable MTs and looked at adjusting these with our covariate PRS, does this make any difference?

```{r}
#First extract the bug names by removing any part of the string after "_res_" in the trait column 
h2out[["h2bugs"]] <- h2out[["h2bugs"]] %>%
  mutate(bug = gsub("_allprs_res", "", trait),
         bug = gsub("_sigprs_res", "", bug))

#Plot estimates from h2out[["dircons"]] so we can see the adjusted and unadjusted h2 estimates, if ends with 5e.08, these are the adjusted phenotypes
h2out[["h2bugs"]] %>%
  ggplot(aes(x = h2, y = bug, color = ifelse(grepl("prs", trait), "Adjusted", "Unadjusted"))) +
  geom_point(position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = h2 - 1.96 * se, xmax = h2 + 1.96 * se), position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "red") +
  theme_minimal() +
  labs(title = "Heritability Estimates of Microbiome Traits Before and After Adjustment for Covariate PRS",
       x = "Heritability Estimate",
       y = "Microbiome Trait",
       color = "Adjustment Type") +
  theme(legend.position = "bottom")
```

* Again, no difference in the hertiable estimates. This suggests that the covariates are not driving the heritability of the MTs, or that power is a driving limitation here