---
title: "Host Genome, Microbiome, Covariates"
author: "Alec McKinlay"
date: "`r Sys.Date()`"
output: 
  html_document:
    theme: journal
    toc: yes
    toc_float:
      collapsed: true
---


<style>
  figure {
    text-align: center;
    margin-bottom: 20px; 
  }
  figcaption {
    font-style: italic;
    font-size: 0.9em;
    color: #555;
  }
</style>


```{r, echo = FALSE, message=FALSE, warning=FALSE}

#Set options for the fig sides
knitr::opts_chunk$set(echo = F, message=F, warning=F)

#Set up and read in files
#setwd to 05_Markdowns folder

#Library 
library(tidyverse)
library(data.table)
library(readxl)
library(kableExtra)
library(ggplot2)
library(ggforestplot)
library(DataExplorer)
library(patchwork)
library(forcats)
library(makepipe)
library(tidyr)
library(ggh4x)

#Data path for proj dir
data.path <- "../../../data/PhenoPRS"

#Also load in path to scratch, proj_dir and fgfp data
source("../../parameters/base_dir.R")

#Load the data in, incase we use this
load(file.path(data.path, "data_out/fgfpdata4prs.RData"))

#Now load the data
load(file.path(data.path, "data_out/data_Reg03.1.RData"))

#Now load the data
load(file.path(data.path, "data_out/data_ConsistentAssoc03.2.RData"))

#State filepath to our figures in the markdown
fallonyjpeg <- file.path(data.path, "markdowns/figs/fallonyfig.jpeg")
methodsjpeg <- file.path(data.path, "markdowns/figs/method.jpg")

#Read in excel spread sheet with covariates
gwas_ref <- read.csv(file.path(data.path, "phenogwas_map.csv"), header = T, skip = 1)

#Load in Genetic cor and h2 adjust results
genetic_cor <- read.table(file.path(data.path, "greml/greml_out/h2out_cor_dircons/clean_out.txt"), header = T,
                          sep = "\t")

obs_prs_cor_filt <- read.table(file = file.path(data.path, "greml/greml_out/h2out_cor_dircons/clean_out_withobsprs.txt"), header = T, sep = "\t")


### Load in beta coefficents from the two GWAS
load(file.path(data.path, "data_out/filtered_bug_sumstats_05.1.RData"))


#Now load in the heritability adjustment results
h2out <- list()
for(i in c("dircons", "h2bugs")){
  h2out[[i]] <- read.table(file.path(data.path, paste0("greml/greml_out/h2_out/", i, "/cleanh2est.txt")), header = T, sep = "\t")
}


#Figs
MR_assumptions <- file.path(data.path, "markdowns/figs/MR_assump.png")
h2_confound <- file.path(data.path, "markdowns/figs/h2_confound.png")
fallonyjpeg <- file.path(data.path, "markdowns/figs/fallonyfig.jpeg")
manplot1 <- file.path(data.path, "data_out/manhattan/manhattan_fgfp/yengo_bmi_G_unclassified_P_Firmicutes_RNT_manhattan_plot2.png")

manplot2 <- file.path(data.path, "data_out/manhattan/manhattan_fgfp/Standing_height_F_Ruminococcaceae_RNT_manhattan_plot2.png")

manplot3 <- file.path(data.path, "data_out/manhattan/manhattan_fgfp/Alanine_aminotransferase_levels_G_Roseburia_RNT_manhattan_plot2.png")

#Map
naming_map <- obsest_prsest %>% select(term, covar_name, term_prs, prs) %>% unique()

```

# Introduction

## Replication crisis in Gut Microbiome GWAS's

Previous chapters aimed to quantify associations between the host genome and human gut microbiome both from published literature and through generation of novel polygenic scores (PRSs) in FGFP, with limited success. Published literature suggest most signals between the genome and derived Gut Microbiome Traits (GMTs) are unable to replicate across study (even as power for these studies increase), whilst polygenic scores generated using the GMT as the primary phenotype show no evidence of an association with the trait the PRS has been trained to proxy.

Despite these inconsistencies, what is clear is that the GMTs derived in various cohorts show evidence for both SNP-based and twin-based heritabiliy. One hypothesis explaining these patterns is that the GWAS signals and observed heritability may be inflated by associations between the gut microbiome and other covariables of the microbiome (e.g. BMI / diet / various metabolite levels), which are themselves heritable. This has been demonstarted in the literature by the one association replicated across many cohorts between the LCT region and relative abundance of the Bifidobacterium genus, as association dependent on milk consumption of the host. Here it is shown that rather than a direct mechanism of associaiton between the genome and bacteria, we see a complex interaction arise between the host genome, diet and gut microbiome.

If all GMT GWAS signals are dependent on complex interactions / indirect mechanisms with other covariables associated to the host, the use of established epidemiological causal inference methods (i.e. Mendelian randomisation) using GMTs as the exposure will potentially be biased through heritable confounding. This could could be one explanation for the number of microbiome MR studies reporting "signal" for a huge number of traits

## Mendelian randomisation

Mendelian randomisation (MR) is a method that uses genetic variants as instrumental variables to estimate the causal effect of an exposure on an outcome. The method is based on the principle that genetic variants are randomly assigned at conception and are therefore not subject to confounding or reverse causation. This allows for the estimation of causal effects in observational data, where confounding and reverse causation are common. For an MR study to be unbiased, the genetic variants used as instruments must satisfy three key assumptions: 

1. The genetic variants are associated with the exposure of interest. 
2. The genetic variants are not associated with any confounders of the exposure-outcome relationship. 
3. The genetic variants are only associated with the outcome through the exposure of interest.

<figure>
  <h4><strong>Figure 1:</strong> Diagramatic overview of MR assumptions</h4>
  <img src="`r MR_assumptions`" width="500" alt="Mendelian randomisation assumptions" />
  <figcaption>For an MR estimand to be valid its three core assumptions should hold, as detailed above.</figcaption>
</figure>


## Heritable Confounding

The complex polygenic nature of traits, leads to associations discovered through GWASs that are likely to arise due to complex phenotypic interaction networks. This mediation is of particular concern in MR if phenotypic pathways open up between the confounder and outcome of interest, independent of the specified primary exposure (in our case GMTs). The figure below highlights some of these potential biases

<figure>
  <h4><strong>Figure 2:</strong> Potential bias through heritable confounding </h4>
  <img src="`r h2_confound`" width="500" alt="Heritable confounding bias" />
  <figcaption>The figure depicts two mechanisms of bias through heritable confounding. G represents genetic variation, X represents our microbiome traits, C are our covariables of the microbiome and Y is the outcome of interest. The left shows situation where some of our instruments are biased through horizontal pleiotropy, whereas the diagram on the right shows a situation where all our instruments are biased and we have complete mispecification of primary phenotype. </figcaption>
</figure>


## Overview

In attempt to understand the heritablity of the microbiome and the potential for heritable confounding bias in downstream MR analyses, we have ran a series of analyses in the FGFP cohort, using genetic, microbiome and phenome data measured on the individuals. 

# Methods

Our methods are as follows:

1. **Selection of Microbiome Covariables**: Using previous results from a previous study we will first identify those phenotypes observed to explain the most variation in microbiome diversity between individuals. 
2. **PRS Generation of covariables**: Suitable candidates for PRS generation will be selected from our covariables of interested by generating PRSs for each of the covariable (where an appropriate GWAS is available) through genome-wide significant SNPs reported on the trait of interest. Those PRSs that explain a significant proportion of the variance in the covariable will be taken forward for further analysis. 
3. **Observational Analysis**: We will then regress the covariable PRSs on the microbiome traits to identify any observational associations between the covariables and the microbiome traits. This will give us an idea of the extent to which the covariables are associated with the microbiome traits, and suitable candidates for relationships we believe to be driving microbiome heritablity estimates. 
4. **PRS Regression**: We will then regress the covariable PRSs on the microbiome traits to identify consistent betas between the observational and genetic regression analyses. 
5. **Genetic Correlation**: We will use the GREML to estimate the genetic correlation between the microbiome traits and the covariates of interest, for those relationships with suggestive observational results (beta p_val < 0.05) and directionality agreement with the PRS estimation.
6. **Recalculation of Microbiome heritable estimates, adjusting for covariate PRSs**: For our candidate relationships, identified from the previous steps - we can now recalculate the heritability of these GMTs adjusting for the PRS downstream analyses has shown evidence of a strong observational and genetic overlap - to identify if this is driving the heritablity observed from the GMT.
7. **Investigation into Microbiome Summary Staistics**: For our candidate relationships, identified from the previous steps - we can now investigate the microbiome summary statistics to see if the SNPs identified in the covariate PRSs are showing up in the microbiome summary statistics, and if these beta coefficents are correlated.


# Results



## Select Covariates of interest

A recent study (Falony et al) used FGFP data to discern which covariables were driving variation in beta diversity. The figure below shows the covariables of the microbiome that were identified as driving variation in beta diversity.

<figure>
  <h4><strong>Figure 3:</strong> Covariables of the Microbiome</h4>
  <img src="`r fallonyjpeg`" width="500" alt="Covariables of the Microbiome" />
  <figcaption>The figure is taken from the falony et al study, which used FGFP data to discern which covariables were driving variation in beta diversity, these are our candidate covariates for downstream analysis</figcaption>
</figure>


To decide which covariables we were interest in taking forward for our analysis, we first decided if there was an appropriate GWAS availble for the covariable of interest using OpenGWAS and the GWASCatalog. We then generated PRSs for each of the covariables (where an appropriate GWAS was available) through genome-wide significant SNPs reported on the trait of interest. Those PRSs that explained a significant proportion of the variance in the covariable were taken forward as our candidate covariables. The following table shows those covariables that were chosen, along with variance explained in the trait observationally by the polygenic score, and the pvalue of the beta coeffiencent (from the regression of covariable PRS onto the covariable in FGFP).

```{r}
## Display the traits we are actually working with

#First make nice to look at
data4prs$phenos2use <- data4prs$phenos2use %>% 
  mutate(covar_name = case_when(
    pheno == "GPT_UL" ~ "Alannine_aminotransferase",
    pheno == "CK_UL" ~ "Creatine_kinase",
    pheno == "Gamma.GT_UL" ~ "Gamma_glutamyltransferase",
    pheno == "HDL.chol_mgdL" ~ "HDL",
    pheno == "Hemoglobine_gdL" ~ "Hemoglobine",
    pheno == "hip_circumference" ~ "Hip_circumference",
    pheno == "MCHC_g_dL" ~ "Mean_corp_hem_conc",
    pheno == "RBC_milj_mm3" ~ "Red_Blood_Cells",
    pheno == "Creatinine_mgdL" ~ "Creatinine",
    pheno == "Ureum_mgdL" ~ "Urea",
    pheno == "Urinezuur_mgdL" ~ "Uric_acid",
    pheno == "Height" ~ "Height",
    pheno == "Triglyceriden_mgdL" ~ "Triglycerides",
    TRUE ~ pheno  # Keep all other values unchanged
  ))

data4prs$r2ofcovars$term2 <- gsub(".Pt_5e.08", "", data4prs$r2ofcovars$term)

viz_df <- merge(data4prs$phenos2use, data4prs$r2ofcovars, by.x="prs", by.y = "term2") %>%
  select(covar_name, `partial_r2(m)[prs]`, p.value)

colnames(viz_df) <- c("Covariate", "Partial R2", "Beta_pval") 

```

```{r}
viz_df %>%
  mutate(`Partial R2` = scales::percent(`Partial R2`, accuracy = 0.01),
         Beta_pval = format(Beta_pval, scientific = TRUE, digits = 2)) %>%
  kbl(caption = "Table 1: Summary of covariables chosen for downstream analyses, along with variance explained by the trained PRS using beta weights from an independent GWAS") %>%
  kable_styling()
```

### Selection of Microbiome traits in FGFP

The gut microbiome was sampled and measured for a subset of the FGFP cohort using 16S rRNA sequencing of stool samples. These were then processed into both continuous and binary traits suitable for regression-based analysis as detailed in the Hughes et al FGFP GWAS. For our analysis here, we have focused on the rank normal transformed (RNT) microbiome traits, as these were suggested to be the most appropriate for normality correction. An overview of these traits are highlighted below

```{r}
#Work out how many of each category there are of the microbiome traits
taxa_group <- filter(obs_est, term == "GPT_UL")["group"]

#Make factor so levels don't change
taxa_group$group <- factor(taxa_group$group, levels = unique(taxa_group$group))

# Create a bar plot using ggplot2
ggplot(taxa_group, aes(x = group)) +
  geom_bar(fill = "skyblue") +   # Bar chart with color
  geom_text(stat = 'count', aes(label = ..count..), vjust = -0.5) +   # Bar chart with color
  labs(title = "Number of Microbiome Outcomes by Taxa Group", 
       x = "Taxa Group", 
       y = "Count of Outcomes") + 
  theme_minimal() +              
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
```

### Number of traits
We are therefore looking at a total of `r nrow(taxa_group)` microbiome traits and `r nrow(viz_df)` covariates, meaning a total of `r nrow(taxa_group) * nrow(viz_df)` relationships will be explored in this pipeline

## Observational Analysis in FGFP

First we can run the observational analysis of all these relationships, regressing our covariate onto the RNT microbiome traits in FGFP. 

Due to the large number of relationships considered, forest plots were far too busy to display data. Instead these relationships can be visualised through heat maps, where the covariables are plotted on the x axis and microbiome traits are plotted on the y axis

```{r, fig.width=14, fig.height=14}
#Make factor so levels don't change
obs_est$group <- factor(obs_est$group, levels = unique(obs_est$group))

#ggforestplot::forestplot(
#  df = obs_est,
#  name = mt,
#  estimate = estimate,
#  se = std.error,
#  pvalue = p.value,
#  colour = term,
#  xlab = "SD increase in Rank normal transformed Microbial trait per 1-SD increase of the covariate #in FGFP",
#  title = "Regression coefficents from Observational analyses of Covariate regressed on RNT #Microbial Traits in FGFP") +
#  facet_wrap(
#    facets = ~group,
#    ncol = 3,  # Number of columns
#    nrow = 2,  # Number of rows
#    scales = "free_y",
#  )

### Heat Map

# Assuming obs_est is your dataframe and has columns: term, mt, p.value, group
obs_est %>%
  # Create a binary column for significance: 1 if p < 0.05, else 0
  mutate(sig = ifelse(p.value < 0.05 & estimate > 0, 1, ifelse(p.value < 0.05 & estimate < 0, 2, 0))) %>%
  
  # Reorder the x-axis based on the number of significant results (descending order)
  group_by(covar_name) %>%
  mutate(sig_count = sum(p.value < 0.05)) %>%
  ungroup() %>%
  mutate(covar_name = reorder(covar_name, -sig_count)) %>%  # Reorder based on significant results
  
  # Create the heatmap
  ggplot(aes(x = covar_name, y = mt, fill = factor(sig))) +  # Convert sig to a factor for discrete color mapping
  geom_tile(color = "grey80", size = 0.5) +  # Add borders to tiles
  
  # Use manual color scale for binary significance: white for 0, blue for 1
  scale_fill_manual(values = c("0" = "white", "1" = "darkgreen", "2" = "purple"),
                    name = "Significance",
                    labels = c("0" = "No Observational Signal (p > 0.05)",
                               "1" = "Positive Observational Estimate (p < 0.05)",
                               "2" = "Negative Observational Estimate (p < 0.05)")) +
  
  # Facet by group on the y-axis to add headers
  facet_grid(group ~ ., scales = "free", space = "free") +  # Grouped by 'group' column, free scaling
  
  # Labels and formatting
  labs(title = "Observational associations between microbiome traits and covariates of interest, indicating where there is evidence of singal",
       x = "Covariate",
       y = "Microbiome Traits") +
  
  # Theme for improved aesthetics
  theme_minimal(base_size = 12) +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1, size = 11),  # Rotate x-axis labels
    axis.text.y = element_text(size = 11),  # Larger y-axis labels
    axis.title = element_text(size = 13, face = "bold"),  # Bold axis titles
    plot.title = element_text(hjust = 0.5, size = 16, face = "bold"),  # Center and bold title
    plot.subtitle = element_text(hjust = 0.5, size = 12),  # Center subtitle
    strip.text.y = element_text(size = 12, face = "bold", angle = 0),  # Bold group labels on y-axis
    panel.grid = element_blank()  # Remove gridlines
  )

```

The plot has been ordered from left to right with the covariable which shows the highest number of signals with the microbiome traits. 

We can also visualise the amount of signal through a bar chart. This bar chart shows the number of observational regression with evidence of signal (p < 0.05) against those with no signal (p > 0.05) for each covariate - GMT relationship.

```{r, fig.width=14, fig.height=14}
# Bar chart
# Assuming obs_est has columns: covar_name, mt, p.value, group
obs_est %>%
  # Create a binary variable for significance (1 = significant, 0 = not significant)
  mutate(sig = ifelse(p.value < 0.05, 1, 0)) %>%
  
  # Group by covariate and significance, then count the number of results in each group
  group_by(covar_name, sig) %>%
  summarise(count = n()) %>%
  ungroup() %>%
  
  # Reorder the x-axis (covar_name) based on the total number of significant results (descending)
  group_by(covar_name) %>%
  mutate(total_sig = sum(ifelse(sig == 1, count, 0))) %>%
  ungroup() %>%
  mutate(covar_name = reorder(covar_name, -total_sig)) %>%
  
  # Create the bar plot
  ggplot(aes(x = covar_name, y = count, fill = factor(sig))) +
  geom_bar(stat = "identity", position = "dodge", color = "black", size = 0.3) +  # Dodged bar chart
  
  # Define the color scale for significant vs non-significant bars
  scale_fill_manual(values = c("0" = "lightgray", "1" = "blue"), 
                    labels = c("No Signal", "Signal (p < 0.05)"), 
                    name = "Association") +
  
  # Add labels and titles
  labs(title = "Count of suggestive signal from observational regressions",
       x = "Covariate",
       y = "Count of Outcomes") +
  
  # Improve theme aesthetics
  theme_minimal(base_size = 12) +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1, size = 11),  # Tilt x-axis labels
    axis.text.y = element_text(size = 11),  # Larger y-axis labels
    axis.title = element_text(size = 13, face = "bold"),  # Bold axis titles
    plot.title = element_text(hjust = 0.5, size = 16, face = "bold"),  # Center and bold title
    legend.position = "top"  # Move legend to top for better layout
  )
```

## Covariate PRS regression on GMTs

For those relationships with signal, we can now run the PRS regression on the GMTs to identify consistent betas between the observational and genetic regression analyses. This can also be visualised in a heat map, where we will colour in the tiles based on the agreement between the observational and genetic regression analyses. Red tile indicates disagreement between the observational and PRS regression, green indicates agreeement, and white indicates no evidence of an observational assocaition

```{r, fig.width=14, fig.height=14}
# Lets revisit our heatmap, but this time with observational agreeement
#Make factor so levels don't change
obsest_prsest$group <- factor(obsest_prsest$group, levels = levels(obs_est$group))

obsest_prsest %>%
  # Update the sig column with descriptive labels
  mutate(sig = ifelse(p.value_obs < 0.05 & sign(estimate_obs) == sign(estimate_prs), "Beta_Agree", 
                      ifelse(p.value_obs < 0.05, "Beta_Disagree", "No_Obs_Signal"))) %>%
  
  # Reorder the x-axis based on the number of significant results (descending order)
  group_by(covar_name) %>%
  mutate(sig_count = sum(sig != "No_Obs_Signal")) %>%  # Counting only the significant results
  ungroup() %>%
  mutate(covar_name = reorder(covar_name, -sig_count)) %>%
  
  # Create the heatmap
  ggplot(aes(x = covar_name, y = mt, fill = sig)) +  # Use the 'sig' column directly for fill
  geom_tile(color = "grey80", size = 0.5) +  # Add borders to tiles
  
  # Custom color scheme: white for No Signal, red for Disagreement, blue for Agreement
  scale_fill_manual(values = c("No_Obs_Signal" = "white", "Beta_Disagree" = "red", "Beta_Agree" = "green"),
                    name = "Significance",
                    labels = c("No_Obs_Signal" = "No Observational Signal",
                               "Beta_Disagree" = "Beta Disagree",
                               "Beta_Agree" = "Beta Agree")) +
  
  # Facet by group on the y-axis to add headers
  facet_grid(group ~ ., scales = "free", space = "free") +  # Grouped by 'group' column, free scaling
  
  # Labels and formatting
  labs(title = "Beta Directionality Agreement Between Observational and PRS Regressions",
       subtitle = "Blue: Beta Agreement, Red: Beta Disagreement, White: No Observational Signal",
       x = "Observational Covariate",
       y = "Microbiome Traits") +
  
  # Theme for improved aesthetics
  theme_minimal(base_size = 12) +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1, size = 11),  # Rotate x-axis labels
    axis.text.y = element_text(size = 11),  # Larger y-axis labels
    axis.title = element_text(size = 13, face = "bold"),  # Bold axis titles
    plot.title = element_text(hjust = 0.5, size = 16, face = "bold"),  # Center and bold title
    plot.subtitle = element_text(hjust = 0.5, size = 12),  # Center subtitle
    strip.text.y = element_text(size = 12, face = "bold", angle = 0),  # Bold group labels on y-axis
    panel.grid = element_blank()  # Remove gridlines
  )

```

We can again visualise this as a bar chart to get a better idea of the number of relationships where the observational and PRS regression agree or disagree.


```{r}
## Lets also make a bar chart so we can see results of this more clearly
obsest_prsest %>%
  # Create a binary variable for significance (1 = significant, 0 = not significant)
  mutate(sig = ifelse(p.value_obs < 0.05 & sign(estimate_obs) == sign(estimate_prs), "Beta_Agree", 
                      ifelse(p.value_obs < 0.05, "Beta_Disagree", "No_Obs_Signal"))) %>%
  
  # Filter out 'No_Obs_Signal' to focus only on significant results
  filter(sig != "No_Obs_Signal") %>%
  
  # Group by covariate and significance, then count the number of results in each group
  group_by(covar_name, sig) %>%
  summarise(count = n()) %>%
  ungroup() %>%
  
  # Reorder the x-axis (covar_name) based on the total number of significant results (descending)
  group_by(covar_name) %>%
  mutate(total_sig = sum(count)) %>%
  ungroup() %>%
  mutate(covar_name = reorder(covar_name, -total_sig)) %>%
  
  # Create the bar plot
  ggplot(aes(x = covar_name, y = count, fill = sig)) +
  geom_bar(stat = "identity", position = "dodge", color = "black", size = 0.3) +  # Dodged bar chart
  
  # Define the color scale for agreement/disagreement bars
  scale_fill_manual(values = c("Beta_Disagree" = "red", "Beta_Agree" = "green"), 
                    labels = c("Beta Agree", "Beta Disagree"), 
                    name = "Association") +
  
  # Add labels and titles
  labs(title = "Bar chart plotting agreement vs disagreement of beta directionality",
       x = "Covariate",
       y = "Count of Outcomes") +
  
  # Improve theme aesthetics
  theme_minimal(base_size = 12) +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1, size = 11),  # Tilt x-axis labels
    axis.text.y = element_text(size = 11),  # Larger y-axis labels
    axis.title = element_text(size = 13, face = "bold"),  # Bold axis titles
    plot.title = element_text(hjust = 0.5, size = 16, face = "bold"),  # Center and bold title
    legend.title = element_text(size = 12),
    legend.text = element_text(size = 11)
  )

```

### Selection of candidate covariate - GMT relationships

Now we want to find those relationships where we have both observational and PRS agreement, and pvalue < 0.05 for both regression analysis. These will be our strongest candidate relationships to further investigate summary statistics of the GMTs and covariables of interest.

We can visualise these forest plots below of our correlated phenotypes.

```{r, fig.width=14, fig.height=14}
# Make a forest plot of the candidiate_bugs
#We can also look at those with sig observational results
#First reshape the data
obsest_prsest_long <- candidate_bugs %>%
  pivot_longer(cols = c(estimate_obs, estimate_prs), 
               names_to = "type", 
               values_to = "estimate") %>%
  pivot_longer(cols = c(std.error_obs, std.error_prs), 
               names_to = "error_type", 
               values_to = "std.error") %>%
  pivot_longer(cols = c(p.value_obs, p.value_prs), 
               names_to = "p_value_type", 
               values_to = "p.value") %>%
  filter((type == "estimate_obs" & error_type == "std.error_obs" & p_value_type == "p.value_obs") |
         (type == "estimate_prs" & error_type == "std.error_prs" & p_value_type == "p.value_prs")) %>%
  mutate(type = recode(type, estimate_obs = "Observational", estimate_prs = "PRS"))



#Forest plot for the observational and PRS estimate
ggforestplot::forestplot(
  df = filter(obsest_prsest_long, group != "Diversity"),
  name = mt,
  estimate = estimate,
  se = std.error,
  pvalue = p.value,
  colour = type,
  xlab = "SD increase in Rank normal transformed Microbial trait per 1-SD increase in polygenic loading of the covariate / the measure covariate",
  title = "Selected Covariate-Microbiome relationships with strongest evidence of observational and PRS signal") +
  ggforce::facet_col(
    facets = ~covar_name,
    scales = "free_y"
  )


```


## Genetic Correlation

For completeness we can also look at the genetic correlation between our candidate relationships using GREML. Again, we can filter to those relationships whereby the directionality of the genetic correlation agrees in the beta directionality of the observational and PRS regression.

Many of the genetic correlations were unable to run for many of the GMTs as the analysis was too underpowered to detect anything. This is likely due to the small sample size of the FGFP cohort. 

```{r, fig.width=14, fig.height=14}
# Combine the observational and PRS estimates for plotting
plot_data <- obs_prs_cor_filt %>%
  select(term, mt, estimate_obs, std.error_obs, estimate_prs, std.error_prs) %>%
  pivot_longer(cols = c(estimate_obs, estimate_prs),
               names_to = "type",
               values_to = "estimate") %>%
  mutate(std.error = ifelse(type == "estimate_obs", std.error_obs, std.error_prs),
         type = ifelse(type == "estimate_obs", "Observational", "PRS"))

# Create a separate data frame for the genetic correlation estimates
rg_data <- obs_prs_cor_filt %>%
  select(term, mt, rg, rg_se) %>%
  rename(estimate = rg, std.error = rg_se) %>%
  mutate(type = "Genetic Correlation")

# Combine the data frames for plotting
combined_data <- bind_rows(plot_data, rg_data)

# Filter out mts that are not relevant for specific terms
filtered_data <- combined_data %>%
  group_by(term, type) %>%
  filter(!is.na(estimate)) %>%
  ungroup()

# Reorder the type factor to ensure the correct order in the plot
filtered_data$type <- factor(filtered_data$type, levels = c("Observational", "PRS", "Genetic Correlation"))

# Create the forest plot with facets
ggplot(filtered_data, aes(x = estimate, y = mt, color = type)) +
  geom_point(position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = estimate - 1.96 * std.error,
                    xmax = estimate + 1.96 * std.error),
                width = 0.2, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "red") +  # Add dashed red line at x = 0
  facet_grid(term ~ type, scales = "free", switch = "y") +
  theme_minimal() +
  labs(title = "Forest Plot of Observational, PRS Betas, and Genetic Correlation",
       x = "Estimate",
       y = "Microbiome Trait",
       color = "Estimate Type") +
  theme(legend.position = "bottom",
        strip.text.y = element_text(angle = 0),  # Adjust angle for term labels
        panel.grid.minor = element_blank())  # Remove minor grid lines
```

## Adjusted h2 of GMTs

We not have a handful of relationships where the observational and PRS regression agree, and the genetic correlation is in the same direction as the beta coefficent. 

Our first planned analysis is to recalculate the heritability of these GMTs adjusting for the PRS of the covariable shown to be correlated with the GMT of interest. This would show evidence of the heritablity of microbiome traits being predmonently driven by associations with the covariable.

```{r}
#First extract the bug names by removing any part of the string after "_res_" in the trait column 
h2out[["dircons"]] <- h2out[["dircons"]] %>%
  mutate(bug = gsub("_res_.*", "", trait))

#Plot estimates from h2out[["dircons"]] so we can see the adjusted and unadjusted h2 estimates, if ends with 5e.08, these are the adjusted phenotypes
h2out[["dircons"]] %>%
  ggplot(aes(x = h2, y = bug, color = ifelse(grepl("5e.08", trait), "Adjusted", "Unadjusted"))) +
  geom_point(position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = h2 - 1.96 * se, xmax = h2 + 1.96 * se), position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "red") +
  theme_minimal() +
  labs(title = "Heritability Estimates of Microbiome Traits Before and After Adjustment for Covariate PRS",
       x = "Heritability Estimate",
       y = "Microbiome Trait",
       color = "Adjustment Type") +
  theme(legend.position = "bottom")
```

We can see no difference between the adjusted and unadjusted heritability estimates, power however is likely to be a huge limtitng factor.

## Comparison of GMT and Covariable Summary statisitcs

### GWAS beta agreement

We can also further investigate the microbiome summary statistics to see if the SNPs identified in the covariate PRSs are showing up in the microbiome summary statistics, and if these beta coefficents are correlated. 

First we can see if the beta coefficents of the covariate and the bug are correlated, by taking those SNPs used to generate the covariate instrument (p < 5e-8 in the covariate discovery GWAS) and regress them against the same SNP in the FGFP bug GWAS. If correlated, it would suggest that the bug GWAS may be capturing the covariate potential leading to bias downstream MR analysis.

To illustrate this point we will select a few examples from our SNP-Covar relationships.

```{r}

file = file.path(data.path, "data_out/BetaCor_05.3.RData")

beta_cor <- mutate(lm_out,
                   mt = paste0(bug, "Res")) %>%
  merge(unique(obsest_prsest_long[,c("term", "mt", "group", "prs", "covar_name")]), 
        by.x = c("covar", "mt") ,
        by.y = c("prs", "mt")) %>% 
  select(-term.x) %>%
  rename(term = term.y) %>%
  mutate(type = "BetaGWAS_Cor")
        
#Plot with PRS and Observational
obsest_prest_betacor_long <- bind_rows(obsest_prsest_long, beta_cor)


#Forest plot for the observational and PRS estimate
ggforestplot::forestplot(
  df = obsest_prest_betacor_long,
  name = mt,
  estimate = estimate,
  se = std.error,
  pvalue = p.value,
  colour = type,
  xlab = "Beta from Observation / PRS / Beta_GWAS regressions",
  title = "Candidate Covariate-Microbiome relationships") +
  ggforce::facet_col(
    facets = ~covar_name,
    scales = "free_y"
  )


#Exemplar relationships to highlight
#G_Aestuariispira and triglycerides
#G_unclassified_O_Clostridiales and Uric acid

# Fit linear model and extract p-value
model <- lm(beta.outcome ~ beta.exposure, data = harm_data_all[["triglycerides"]][["G_Aestuariispira_RNT"]])
p_value <- summary(model)$coefficients[2, 4] %>% format(scientific = TRUE, digits = 3)

scatter1 <- ggplot(harm_data_all[["triglycerides"]][["G_Aestuariispira_RNT"]], aes(beta.exposure, beta.outcome)) +
    geom_errorbar(aes(ymax = beta.outcome + 1.96 * se.outcome, ymin = beta.outcome - 1.96 * se.outcome), 
                  width = 0, alpha = 0.5) +  # Semi-transparent error bars
    geom_errorbar(aes(xmax = beta.exposure + 1.96 * se.exposure, xmin = beta.exposure - 1.96 * se.exposure), 
                  width = 0, alpha = 0.5) +
    geom_point(shape = 21, color = "black", fill = "steelblue", size = 3) +  # Points with black outline
    geom_smooth(method = "lm", color = "firebrick", linetype = "solid", size = 1) +
    labs(title = "Beta Coefficient Correlation for Triglyceride SNPs\nbetween G_Aestuariispira GWAS and Triglycerides GWAS",
         x = "Triglyceride SNP Betas from G_Aestuariispira GWAS",
         y = "Triglyceride SNP Betas from Triglycerides GWAS") +
    # Annotate p-value on the plot
    annotate("text", x = Inf, y = Inf, label = paste("Slope p-value:", p_value), 
             hjust = 1.1, vjust = 1.1, size = 3, color = "black") +
    theme_classic() +
    theme(
        plot.title = element_text(size = 14, face = "bold", hjust = 0.5),
        axis.title = element_text(size = 12, face = "bold"),
        axis.text = element_text(size = 10),
        plot.margin = unit(c(1, 1, 1, 1), "lines")
    )

model <- lm(beta.outcome ~ beta.exposure, data = harm_data_all[["Serum_uric_acid_levels"]][["G_unclassified_O_Clostridiales_RNT"]])
p_value <- summary(model)$coefficients[2, 4] %>% format(scientific = TRUE, digits = 3)

scatter2 <- ggplot(harm_data_all[["Serum_uric_acid_levels"]][["G_unclassified_O_Clostridiales_RNT"]], aes(beta.exposure, beta.outcome)) +
    geom_errorbar(aes(ymax = beta.outcome + 1.96 * se.outcome, ymin = beta.outcome - 1.96 * se.outcome), 
                  width = 0, alpha = 0.5) +  # Semi-transparent error bars
    geom_errorbar(aes(xmax = beta.exposure + 1.96 * se.exposure, xmin = beta.exposure - 1.96 * se.exposure), 
                  width = 0, alpha = 0.5) +
    geom_point(shape = 21, color = "black", fill = "steelblue", size = 3) +  # Points with black outline
    geom_smooth(method = "lm", color = "firebrick", linetype = "solid", size = 1) +
    labs(title = "Beta Coefficient Correlation for Uric Acid SNPs\nbetween G_unclassified_O_Clostridiales GWAS and Uric Acid GWAS",
         x = "Uric Acid SNP Betas from G_unclassified_O_Clostridiales GWAS",
         y = "Uric Acid SNP Betas from Uric Acid GWAS") +
    # Annotate p-value on the plot
    annotate("text", x = Inf, y = Inf, label = paste("Slope p-value:", p_value), 
             hjust = 1.1, vjust = 1.1, size = 3, color = "black") +
    theme_classic() +
    theme(
        plot.title = element_text(size = 14, face = "bold", hjust = 0.5),
        axis.title = element_text(size = 12, face = "bold"),
        axis.text = element_text(size = 10),
        plot.margin = unit(c(1, 1, 1, 1), "lines")
    )


forestplot <- ggforestplot::forestplot(
  df = filter(obsest_prsest_long, 
              (mt == "G_Aestuariispira_RNTRes" & prs == "triglycerides") | 
              (mt == "G_unclassified_O_Clostridiales_RNTRes" & prs == "Serum_uric_acid_levels")),
  name = mt,
  estimate = estimate,
  se = std.error,
  pvalue = p.value,
  colour = type,
  xlab = "SD increase in Rank normal transformed Microbial trait per 1-SD increase in polygenic loading of the covariate / the measure covariate in FGFP",
  title = ("PRS and Observational Forest plot")) +
  ggforce::facet_row(
    facets = ~covar_name,
    scales = "free_y"
  ) +
  # Apply theme adjustments
  theme_classic() +
  theme(
    plot.title = element_text(size = 14, face = "bold", hjust = 0.5),  # Title style
    axis.title.x = element_text(size = 12, face = "bold"),  # X-axis title style
    axis.text.x = element_text(size = 10),  # X-axis tick label style
    strip.text = element_text(size = 12, face = "bold"),  # Facet label style
    plot.margin = unit(c(1, 1, 1, 1), "lines")  # Adjust plot margins
  )

forestplot / (scatter1 + scatter2) + 
  plot_layout(heights = c(1, 3)) + plot_annotation(tag_levels = "A")


#Plot the beta coefficents with the forest plot
#Now we can do the same for just the diversity metrics
plot1 <- ggforestplot::forestplot(
  df = filter(obsest_prsest_long, group == "Diversity"),
  name = mt,
  estimate = estimate,
  se = std.error,
  pvalue = p.value,
  colour = type,
  xlab = "SD increase in Rank normal transformed Microbial trait per 1-SD increase in polygenic loading of the covariate / the measure covariate",
  title = "Selected Covariate-Microbiome relationships with strongest evidence of observational and PRS signal") +
  ggforce::facet_row(
    facets = ~covar_name,
    scales = "free_y"
  )


ggplot(harm_data_all[["yengo_bmi"]][["Div_Chao1_RNT"]], aes(beta.exposure, beta.outcome)) +
      geom_point() +
  geom_errorbar(aes(ymax = beta.outcome + 1.96*se.outcome, ymin = beta.outcome - 1.96*se.outcome)) +
      geom_errorbar(aes(xmax = beta.exposure + 1.96*se.exposure, xmin = beta.exposure - 1.96*se.exposure)) +
      geom_smooth(method="lm")
      
  
  
  
  
   +
labs(title = paste0("Beta Coefficent Correlation between ", bug, " and ", covar),
     x = "Beta coefficents from covariate GWAS",
     y = "Beta coefficients from bug GWAS for covariate SNPs") #+
#geom_abline(intercept=0, slope=1)

lm(beta.exposure 

#Lets also focus in on those relationships where we are looking at in MiBioGen





```


When we see a strong positive correlation in our PRS and Observational results, we can see that the beta coefficents from the two GWAS's are positively correlated (e.g. G_Aestuariispira and triglycerides) whereas a negative relationship we then see that the beta coefficents are negatively correlated (e.g. G_unclassified_O_Clostridiales and Uric acid)

Below is the plots showing the correlation between beta coefficents of the bug and covariable

```{r, fig.width=14, fig.height=8}
plot[[1]] + plot[[6]]
```


### Microbiome GWAS signals

We can also observe the microbiome summary statistics to see if the SNPs identified in the covariate PRSs are showing up in the microbiome summary statistics. This will be especially problematic if these covariate SNPs start to creep above the relaxed pvalue selection threshold (1e-5) in the microbiome GWAS as this is the current threshold used in the literature to select instruments for Mendelian radnomsiation analysis

We can colour in SNPs which are shown to be associated with the covariate GWAS on the microbiome manhattan plot from FGFP. As an example we can use the FGFP GWAS for G_unclassified_P_Firmicutes which was able to find signal in this cohort at genome-wide significance level as observed in the manhattan plot. However, results outlined  BMI and this particular GMT as one of the relationships to covary with one another. If we colour in BMI snps (and those SNPs in LD with the BMI snps), we can see that the SNPs associated with BMI are showing up in the microbiome summary statistics:

<figure>
  <h4>Manhattan plot of G_unclassified_P_Firmicutes GWAS results, colouring in those SNPs associated (or in LD with those associated) with BMI</h4>
  <img src="`r manplot1`" width="500" alt="manplot1" />
  <figcaption></figcaption>
</figure>


We can also observe this with several other of our candidate relationship - whereby SNPs associated with the covariate begin to creep into the the PRS instrument for the bug if we relax the threshold

<figure>
  <h4><strong>Figure 1:</strong>Manhattan plot of F_Ruminococcaceae GWAS results, colouring in those SNPs associated (or in LD with those associated) with Height</h4>
  <img src="`r manplot2`" width="500" alt="manplot3" />
  <figcaption></figcaption>
</figure>

<figure>
  <h4><strong>Figure 1:</strong>Manhattan plot of G_Roseburia GWAS results, colouring in those SNPs associated (or in LD with those associated) with AAT levels</h4>
  <img src="`r manplot3`" width="500" alt="manplot3" />
  <figcaption></figcaption>
</figure>


This is particularly problematic when lowering the pvalue threshold for selecting instruments for MR analysis, as the covariate SNPs are showing up in the microbiome summary statistics. This could lead to bias in the MR analysis, if the bug is not on the causal path from SNP -> covar -> outcome.


```{r}
# Lets load in the plots generated from the previous Rscript
load(file.path(data.path, "data_out/mibiogen_manplots.RData"))

#Ones to highlight as examples
#Roseburia, Porphyromonadaceae, Erysipelotrichaceae, Firmicutes

traits4plots <- c("genus.Roseburia.id.2012", 
                  "family.Porphyromonadaceae.id.943",
                  "family.Erysipelotrichaceae.id.2149",
                  "phylum.Firmicutes.id.1672")

plots <- list()

for(i in 1:length(traits4plots)){

mt_mbg <- traits4plots[i]
mt_fgfp <- matched_bugs[matched_bugs$mbgbugs == mt_mbg, "fgfpbugs"]

# Your forest plot code
forestplot_plot <- ggforestplot::forestplot(
  df = filter(obsest_prest_betacor_long, mt == mt_fgfp),
  name = mt,
  estimate = estimate,
  se = std.error,
  pvalue = p.value,
  colour = type,
  xlab = "Beta",
  title = "Beta coefficents from FGFP analysis"
) +
  ggforce::facet_row(
    facets = ~covar_name
  ) + labs(color = "Regression_Type")

plots[[mt_mbg]][[1]] <- forestplot_plot / manplots[[mt_mbg]] + plot_layout(heights = c(1,4))

plots[[mt_mbg]][[2]] <- manplots[[mt_mbg]] + inset_element(forestplot_plot, left = 0, bottom = 0, right = 0.5, top = 0.3)

}

```
